### DISCUSSION TITLE
Halide::BoundaryConditions::repeat_edge: How to repeat left, bottom, right, and top elemets of Buffer instead of only Bottom and right?

### DISCUSSION THREAD

**Discussion by FabianSchuetze** [halide/Halide#8208] [Q&A] [answered]
Thanks for the wonderful library! 

Suppose I have the following buffer:
```
    Buffer<float, 2> input(3, 3);
    input(0, 0) = 0;
    input(1, 0) = 1;
    input(2, 0) = 2;
    input(0, 1) = 3;
    input(1, 1) = 4;
    input(2, 1) = 5;
    input(0, 2) = 6;
    input(1, 2) = 7;
    input(2, 2) = 8;
```
and the following algorithm, with a output that is of size (5,5) (two more rows and cols):
```
bounded_input = Halide::BoundaryConditions::repeat_edge(input); 
```
That generates the following buffer:
```
Store repeat_edge.0(0, 0) = 0.000000
Store repeat_edge.0(1, 0) = 1.000000
Store repeat_edge.0(2, 0) = 2.000000
Store repeat_edge.0(3, 0) = 2.000000
Store repeat_edge.0(4, 0) = 2.000000
Store repeat_edge.0(0, 1) = 3.000000
Store repeat_edge.0(1, 1) = 4.000000
Store repeat_edge.0(2, 1) = 5.000000
Store repeat_edge.0(3, 1) = 5.000000
Store repeat_edge.0(4, 1) = 5.000000
Store repeat_edge.0(0, 2) = 6.000000
Store repeat_edge.0(1, 2) = 7.000000
Store repeat_edge.0(2, 2) = 8.000000
Store repeat_edge.0(3, 2) = 8.000000
Store repeat_edge.0(4, 2) = 8.000000
Store repeat_edge.0(0, 3) = 6.000000
Store repeat_edge.0(1, 3) = 7.000000
Store repeat_edge.0(2, 3) = 8.000000
Store repeat_edge.0(3, 3) = 8.000000
Store repeat_edge.0(4, 3) = 8.000000
Store repeat_edge.0(0, 4) = 6.000000
Store repeat_edge.0(1, 4) = 7.000000
Store repeat_edge.0(2, 4) = 8.000000
Store repeat_edge.0(3, 4) = 8.000000
Store repeat_edge.0(4, 4) = 8.000000
```
It seems that the right and bottom boundary are extended by two columns (rows). I would have expected to repeat the top and bottom rows and left and right columns. For example, the first row of the new buffer would be:
```
Store repeat_edge.0(0, 0) = 0.000000
Store repeat_edge.0(1, 0) = 0.000000
Store repeat_edge.0(2, 0) = 1.000000
Store repeat_edge.0(3, 0) = 2.000000
Store repeat_edge.0(4, 0) = 2.000000
```

I understand that `repeat_edge` also accepts a `Region` argument. Maybe I am not passing that correctly to achieve my intended effect? For reference, I would like to implement Pytorch's [ReplicationPad](https://pytorch.org/docs/stable/generated/torch.nn.ReplicationPad2d.html#torch.nn.ReplicationPad2d) functionality.

**Comment by abadams** [comment, marked as answer]
repeat_edge will pad on all sides. However, unlike ReplicationPad, repeat_edge also preserves the coordinate system in that f(x, y) before = f(x, y) after if (x, y) is in-bounds. So (0, 0) before is still (0, 0) after, and I think that might be what's tripping you up. To match ReplicationPad, try adding a translation:

```
Func f = repeat_edge(input);
Func translate;
translate(x, y) = f(x-1, y-1);
```

**Reply by FabianSchuetze** [reply]
Thanks a lot for the reply and explanations - very much appreciated!
